src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.Websockets.cs

```diff
@@ -0,0 +1,235 @@
+using Microsoft.AspNetCore.HttpSys.Internal;
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Server.IISIntegration
+{
+    /// <summary>
+    /// Represents the websocket portion of the <see cref="IISHttpContext"/>
+    /// </summary>
+    internal partial class IISHttpContext
+    {
+        private bool _wasUpgraded; // Used for detecting repeated upgrades in IISHttpContext
+
+        private IISAwaitable _readWebSocketsOperation;
+        private IISAwaitable _writeWebSocketsOperation;
+        private TaskCompletionSource<object> _upgradeTcs;
+
+        private async Task CheckForUpgrade()
```

Comment:
CheckForUpgrade is a strange name since it's really just starting the bidirectional stream.